File: Disjoint Set/17352번 - 여러분의 다리가 되어 드리겠습니다!/17352.py
import sys
sys.setrecursionlimit(10**7)

input = sys.stdin.readline

n = int(input())
bridges = [tuple(map(int, input().split())) for _ in range(n - 2)]
parents = list(range(n + 1))

def Find(a):
    if a != parents[a]:
        parents[a] = Find(parents[a])
    return parents[a]

def Union(a, b):
    a = Find(a)
    b = Find(b)
    if a != b:
        parents[b] = a

def solve():
    
    for a, b in bridges:
        Union(a, b)
    
    for i in range(2, n + 1):
        if Find(1) != Find(i):
            print(f"1 {i}")
            break

if __name__ == '__main__':
    solve()

File: Disjoint Set/17352번 - 여러분의 다리가 되어 드리겠습니다!/17352.java
import java.io.*;
import java.util.*;

public class Main {
    static int n;
    static List<int[]> bridges = new ArrayList<>();
    static int[] parents;
    
    public static void main(String[] args) throws IOException {
        input();
        solve();
    }
    
    static void input() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        
        n = Integer.parseInt(br.readLine());
        parents = new int[n + 1];
        
        for (int i = 0; i < n - 2; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            bridges.add(new int[]{a, b});
        }
        
        for (int i = 0; i <= n; i++) {
            parents[i] = i;
        }
        
        br.close();
    }
    
    static void solve() {
        for (int[] bridge : bridges) {
            int a = bridge[0];
            int b = bridge[1];
            if (a > b) {
                int tmp = a;
                a = b;
                b = tmp;
            }
            
            union(a, b);
        }
        
        for (int i = 2; i <= n; i++) {
            if (find(1) != find(i)) {
                System.out.println(1 + " " + i);
                break;
            }
        }
    }
    
    static int find(int x) {
        if (x == parents[x]) return x;
        return parents[x] = find(parents[x]);
    }
    
    static void union(int a, int b) {
        a = find(a);
        b = find(b);
        if (a != b) parents[b] = a;
    }
}